x-pack/plugins/security_solution/public/attack_discovery/hooks/use_poll_api.tsx

```diff
@@ -0,0 +1,131 @@
+  connectorId: string;
+}
+
+export const usePollApi = ({
+  http,
+  setApproximateFutureTime,
+  toasts,
+  connectorId,
+}: Props): {
+  status: AttackDiscoveryStatus | null;
+  data: AttackDiscoveryData | null;
+  pollApi: () => void;
+} => {
+  const [status, setStatus] = useState<AttackDiscoveryStatus | null>(null);
+  const [data, setData] = useState<AttackDiscoveryData | null>(null);
+  const currentConnectorId = useRef<string | undefined>(undefined);
+
+  useEffect(() => {
+    currentConnectorId.current = connectorId;
+    return () => {
+      currentConnectorId.current = undefined;
+    };
+  }, [connectorId]);
+
+  const handleResponse = useCallback(
+    (responseData: AttackDiscoveryResponse) => {
+      if (connectorId == null || connectorId === '') {
+        throw new Error('Invalid connector id');
+      }
+      setStatus(responseData.status);
+      setApproximateFutureTime(
+        moment(responseData.updatedAt).add(responseData.averageIntervalMs, 'milliseconds').toDate()
+      );
+      setData({
+        ...responseData,
+        connectorId,
+        attackDiscoveries: responseData.attackDiscoveries.map((attackDiscovery) => ({
+          ...attackDiscovery,
+          id: attackDiscovery.id ?? uuid.v4(),
+          detailsMarkdown: replaceNewlineLiterals(attackDiscovery.detailsMarkdown),
+          entitySummaryMarkdown: replaceNewlineLiterals(attackDiscovery.entitySummaryMarkdown),
+          summaryMarkdown: replaceNewlineLiterals(attackDiscovery.summaryMarkdown),
+        })),
+      });
+    },
+    [connectorId, setApproximateFutureTime]
+  );
+  const pollApi = useCallback(async () => {
+    try {
+      if (connectorId == null || connectorId === '') {
+        throw new Error('Invalid connector id');
+      }
+      // call the internal API to generate attack discoveries:
+      const rawResponse = await http.fetch(
+        `/internal/elastic_assistant/attack_discovery/${connectorId}`,
+        {
+          method: 'GET',
+          version: ELASTIC_AI_ASSISTANT_INTERNAL_API_VERSION,
+        }
+      );
+
+      const parsedResponse = AttackDiscoveryGetResponse.safeParse(rawResponse);
+      if (!parsedResponse.success) {
+        throw new Error('Failed to parse the attack discovery GET response');
+      }
+      if (parsedResponse.data.entryExists === false || parsedResponse.data.data == null) {
+        setStatus(null);
+        setData(null);
+      } else if (
+        parsedResponse.data.data.status === attackDiscoveryStatus.succeeded ||
+        parsedResponse.data.data.status === attackDiscoveryStatus.failed
+      ) {
+        handleResponse(parsedResponse.data.data);
+      } else if (parsedResponse.data.data.status === attackDiscoveryStatus.running) {
+        handleResponse(parsedResponse.data.data);
+        // poll every 3 seconds if attack discovery is running
+        setTimeout(() => {
+          // react is being annoying and the setTimeout is still running after the connectorId changes
```

Comment:
Is this happening because the current timeout isn't canceled when it changes?

Reply:
that worked! thanks https://github.com/elastic/kibana/pull/184949/commits/36ec361f3f70fa0c5ff0809f2a7a30fe941cb25b